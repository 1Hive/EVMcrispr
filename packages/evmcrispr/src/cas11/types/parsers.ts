import type { Parser } from 'arcsecond';

import type { NodeParserState } from '../parsers/utils';

import type { Node } from './ast';

export type NodeParser<T = Node> = Parser<T, string, NodeParserState>;
