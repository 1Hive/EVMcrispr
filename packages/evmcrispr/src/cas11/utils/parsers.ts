import type { Err } from 'arcsecond';

import type { NodeParserState } from '../parsers/utils';

export const buildParserError = (
  { index, error }: Err<string, null> | Err<string, NodeParserState>,
  type: string,
  msg?: string,
): string => {
  const splitRes = error.split('got');
  const wrongValueEncountered =
    splitRes.length === 2 ? splitRes[1].trim() : null;

  const errorMessage = msg
    ? `${msg}${wrongValueEncountered ? `, got ${wrongValueEncountered}` : ''}`
    : error.split('): ')[1];
  return `${type}(col: ${index}): ${errorMessage}`;
};

export const getIncorrectReceivedValue = (errorMsg: string): string => {
  const splitRes = errorMsg.split('got ');

  if (splitRes.length === 2) {
    return `, got ${splitRes[1].trim()}`;
  }

  return '';
};
