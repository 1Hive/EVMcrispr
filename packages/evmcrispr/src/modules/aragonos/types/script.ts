import type { Address } from "viem";

/**
 * A call script.
 */
export interface CallScriptAction {
  /**
   * The action's target.
   */
  to: Address;
  /**
   * The action's calldata.
   */
  data: `0x${string}`;
}
