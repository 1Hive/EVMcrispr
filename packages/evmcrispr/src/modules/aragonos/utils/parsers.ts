import type { PublicClient } from "viem";
import { parseAbiItem } from "viem";

import type { Address } from "../../../types";
import type { ParsedApp, Repo } from "../types";
import { getSystemApp, isSystemApp } from "./interfaces";

export const parseAppArtifactName = (name: string): string => {
  if (!name) {
    return "";
  }
  // Split by the first '.' occurrence only.
  const parsedName = name.split(/\.(.+)/);

  return parsedName.length > 1 ? parsedName[1] : "";
};

const fetchImplementationAddress = (
  appAddress: Address,
  client: PublicClient,
): Promise<Address> => {
  return client.readContract({
    address: appAddress,
    abi: [
      parseAbiItem("function implementation() public view returns (address)"),
    ],
    functionName: "implementation",
  });
};

export const parseApp = async (
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  app: any,
  client: PublicClient,
): Promise<ParsedApp> => {
  const { address, appId, roles, version } = app;
  const { name: repoName, registry } = app.repo || {};
  const { codeAddress, artifact: rawArtifact, contentUri } = version || {};
  let artifact, name;

  if (isSystemApp(appId)) {
    const systemApp = getSystemApp(appId)!;
    artifact = systemApp.artifact;
    name = systemApp.name;
  } else {
    artifact = JSON.parse(rawArtifact ?? null);
    name = repoName;
  }

  return {
    address,
    appId,
    artifact,
    codeAddress:
      codeAddress ?? (await fetchImplementationAddress(address, client)),
    contentUri,
    name,
    registryName: registry?.name,
    roles: (roles as any[]).map((role) => ({
      ...role,
      roleHash: role["hash"] ?? role["roleHash"],
    })),
  };
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseRepo = (repo: any): Repo => {
  const { artifact: rawArtifact, contentUri, codeAddress } = repo.lastVersion;

  return {
    artifact: JSON.parse(rawArtifact),
    contentUri,
    codeAddress,
  };
};
