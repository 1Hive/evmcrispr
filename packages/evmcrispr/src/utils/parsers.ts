import { getSystemApp, isSystemApp } from '.';
import type { ParsedApp, Repo } from '../types';

export const parseAppArtifactName = (name: string): string => {
  if (!name) {
    return '';
  }
  // Split by the first '.' occurrence only.
  const parsedName = name.split(/\.(.+)/);

  return parsedName.length > 1 ? parsedName[1] : '';
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseApp = (app: any): ParsedApp => {
  const { address, appId, implementation, repoName, roles, version } = app;
  const { registry } = app.repo || {};
  const { codeAddress, artifact: rawArtifact, contentUri } = version || {};
  let artifact, name;

  if (isSystemApp(appId)) {
    const systemApp = getSystemApp(appId)!;
    artifact = systemApp.artifact;
    name = systemApp.name;
  } else {
    artifact = JSON.parse(rawArtifact ?? null);
    name = repoName;
  }

  return {
    address,
    appId,
    artifact,
    codeAddress: codeAddress ?? implementation.address,
    contentUri,
    name,
    registryName: registry?.name,
    roles,
  };
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseRepo = (repo: any): Repo => {
  const { artifact: rawArtifact, contentUri, codeAddress } = repo.lastVersion;

  return {
    artifact: JSON.parse(rawArtifact),
    contentUri,
    codeAddress,
  };
};

export const timeUnits: { [key: string]: number } = {
  s: 1,
  m: 60,
  h: 3600,
  d: 86400,
  w: 604800,
  mo: 2592000,
  y: 31536000,
};
