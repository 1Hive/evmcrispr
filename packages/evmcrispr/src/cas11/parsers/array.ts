import type { Parser } from 'arcsecond';
import {
  between,
  char,
  coroutine,
  recursiveParser,
  sequenceOf,
} from 'arcsecond';

import type { ArrayExpressionNode } from '../types';
import { NodeType } from '../types';

import { argumentExpressionParser } from './expression';

import { commaSeparated, optionalWhitespace } from './utils';

export const arrayExpressionParser: Parser<any, string, any> = recursiveParser(
  () =>
    between(sequenceOf([char('['), optionalWhitespace]))(
      sequenceOf([optionalWhitespace, char(']')]),
    )(commaSeparated(argumentExpressionParser)).map(
      (elements: any): ArrayExpressionNode => {
        return {
          type: NodeType.ArrayExpression,
          elements,
        };
      },
    ),
);

export const arrayExpresionParser = recursiveParser(() =>
  coroutine(function* () {
    yield char('[');
    yield optionalWhitespace;
    const elements = yield commaSeparated(argumentExpressionParser);
    yield optionalWhitespace;
    yield char(']');

    return {
      type: NodeType.ArrayExpression,
      elements,
    };
  }),
);
