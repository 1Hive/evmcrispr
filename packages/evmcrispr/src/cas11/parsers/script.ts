import type { Parser } from 'arcsecond';
import { coroutine, toValue } from 'arcsecond';
import type { AST } from 'prettier';

import { deepConsoleLog } from '../../../test/test-helpers/cas11';

import { ASTType } from '../types';
import { commandExpressionParser } from './command';
import { optionalEmptyLines } from './utils';

export const scriptParser: Parser<AST> = coroutine(function* () {
  const lines = yield optionalEmptyLines(commandExpressionParser);
  return {
    type: ASTType.Program,
    body: lines,
  };
});

export const parseScript = (script: string): AST | undefined => {
  const res = scriptParser.run(script);

  deepConsoleLog(res);
  if (res.isError) {
    console.log(res.error);
    return;
  }

  return toValue(res);
};
