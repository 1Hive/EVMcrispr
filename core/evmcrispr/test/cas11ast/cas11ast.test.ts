import type { Case } from '@1hive/evmcrispr-test-common';
import { DAO, DAO2, DAO3 } from '@1hive/evmcrispr-test-common';

import type { Cas11AST } from '../../src/Cas11AST';
import { parseScript } from '../../src/parsers/script';

describe.concurrent('Cas11AST', () => {
  const script = `
    load aragonos as ar
    load giveth as giv

    ar:connect ${DAO.kernel} (
      set $dao1Variable agent
      connect ${DAO2.kernel} (
        set $dao2Variable vault
        install vault:new
      )
    )

    ar:connect ${DAO3} (
      revoke voting token-manager MINT_ROLE
    )

    set $globalScopeVariable "test"
  `;
  let ast: Cas11AST;

  beforeEach(() => {
    ast = parseScript(script).ast;
  });

  describe.concurrent('when fetching a command at a specific line', () => {
    it('should fetch it correctly', () => {
      expect(ast.getCommandAtLine(9)).toMatchSnapshot();
    });

    describe.concurrent.each<Case<number>>([
      {
        title: 'a line higher than the maximum script line',
        value: 30,
      },
      { title: 'an empty line', value: 4 },
      { title: 'a line without a command', value: 10 },
    ])('', ({ title, value }) => {
      it(`should return nothing when given ${title}`, () => {
        expect(ast.getCommandAtLine(value)).to.be.undefined;
      });
    });
  });

  describe.concurrent('when fetching commands until a specific line', () => {
    it('should fetch them correctly', () => {
      expect(ast.getCommandsUntilLine(12)).toMatchSnapshot();
    });

    describe.concurrent(
      'when given a set of global scope command names',
      () => {
        it('should fetch them correctly', () => {
          expect(
            ast.getCommandsUntilLine(9, ['load', 'set']),
          ).toMatchSnapshot();
        });

        it('should fetch them correctly when giving a line higher than the maximum script line', () => {
          expect(
            ast.getCommandsUntilLine(200, ['load', 'set']),
          ).toMatchSnapshot();
        });
      },
    );
  });
});
