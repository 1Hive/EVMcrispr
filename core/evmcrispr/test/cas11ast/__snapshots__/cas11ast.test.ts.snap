// Vitest Snapshot v1

exports[`Cas11AST > when fetching a command at a specific line > should fetch it correctly 1`] = `
{
  "args": [
    {
      "loc": {
        "end": {
          "col": 25,
          "line": 9,
        },
        "start": {
          "col": 16,
          "line": 9,
        },
      },
      "type": "ProbableIdentifier",
      "value": "vault:new",
    },
  ],
  "loc": {
    "end": {
      "col": 25,
      "line": 9,
    },
    "start": {
      "col": 8,
      "line": 9,
    },
  },
  "name": "install",
  "opts": [],
  "type": "CommandExpression",
}
`;

exports[`Cas11AST > when fetching commands until a specific line > should fetch them correctly 1`] = `
[
  {
    "args": [
      {
        "left": {
          "loc": {
            "end": {
              "col": 17,
              "line": 2,
            },
            "start": {
              "col": 9,
              "line": 2,
            },
          },
          "type": "ProbableIdentifier",
          "value": "aragonos",
        },
        "loc": {
          "end": {
            "col": 23,
            "line": 2,
          },
          "start": {
            "col": 9,
            "line": 2,
          },
        },
        "right": {
          "loc": {
            "end": {
              "col": 23,
              "line": 2,
            },
            "start": {
              "col": 21,
              "line": 2,
            },
          },
          "type": "ProbableIdentifier",
          "value": "ar",
        },
        "type": "AsExpression",
      },
    ],
    "loc": {
      "end": {
        "col": 23,
        "line": 2,
      },
      "start": {
        "col": 4,
        "line": 2,
      },
    },
    "name": "load",
    "opts": [],
    "type": "CommandExpression",
  },
  {
    "args": [
      {
        "left": {
          "loc": {
            "end": {
              "col": 15,
              "line": 3,
            },
            "start": {
              "col": 9,
              "line": 3,
            },
          },
          "type": "ProbableIdentifier",
          "value": "giveth",
        },
        "loc": {
          "end": {
            "col": 22,
            "line": 3,
          },
          "start": {
            "col": 9,
            "line": 3,
          },
        },
        "right": {
          "loc": {
            "end": {
              "col": 22,
              "line": 3,
            },
            "start": {
              "col": 19,
              "line": 3,
            },
          },
          "type": "ProbableIdentifier",
          "value": "giv",
        },
        "type": "AsExpression",
      },
    ],
    "loc": {
      "end": {
        "col": 22,
        "line": 3,
      },
      "start": {
        "col": 4,
        "line": 3,
      },
    },
    "name": "load",
    "opts": [],
    "type": "CommandExpression",
  },
  {
    "args": [
      {
        "loc": {
          "end": {
            "col": 57,
            "line": 5,
          },
          "start": {
            "col": 15,
            "line": 5,
          },
        },
        "type": "AddressLiteral",
        "value": "0x1fc7e8d8e4bbbef77a4d035aec189373b52125a8",
      },
      {
        "body": [
          {
            "args": [
              {
                "loc": {
                  "end": {
                    "col": 23,
                    "line": 6,
                  },
                  "start": {
                    "col": 10,
                    "line": 6,
                  },
                },
                "type": "VariableIdentifier",
                "value": "$dao1Variable",
              },
              {
                "loc": {
                  "end": {
                    "col": 29,
                    "line": 6,
                  },
                  "start": {
                    "col": 24,
                    "line": 6,
                  },
                },
                "type": "ProbableIdentifier",
                "value": "agent",
              },
            ],
            "loc": {
              "end": {
                "col": 29,
                "line": 6,
              },
              "start": {
                "col": 6,
                "line": 6,
              },
            },
            "name": "set",
            "opts": [],
            "type": "CommandExpression",
          },
          {
            "args": [
              {
                "loc": {
                  "end": {
                    "col": 56,
                    "line": 7,
                  },
                  "start": {
                    "col": 14,
                    "line": 7,
                  },
                },
                "type": "AddressLiteral",
                "value": "0x8ccbeab14b5ac4a431fffc39f4bec4089020a155",
              },
              {
                "body": [
                  {
                    "args": [
                      {
                        "loc": {
                          "end": {
                            "col": 25,
                            "line": 8,
                          },
                          "start": {
                            "col": 12,
                            "line": 8,
                          },
                        },
                        "type": "VariableIdentifier",
                        "value": "$dao2Variable",
                      },
                      {
                        "loc": {
                          "end": {
                            "col": 31,
                            "line": 8,
                          },
                          "start": {
                            "col": 26,
                            "line": 8,
                          },
                        },
                        "type": "ProbableIdentifier",
                        "value": "vault",
                      },
                    ],
                    "loc": {
                      "end": {
                        "col": 31,
                        "line": 8,
                      },
                      "start": {
                        "col": 8,
                        "line": 8,
                      },
                    },
                    "name": "set",
                    "opts": [],
                    "type": "CommandExpression",
                  },
                  {
                    "args": [
                      {
                        "loc": {
                          "end": {
                            "col": 25,
                            "line": 9,
                          },
                          "start": {
                            "col": 16,
                            "line": 9,
                          },
                        },
                        "type": "ProbableIdentifier",
                        "value": "vault:new",
                      },
                    ],
                    "loc": {
                      "end": {
                        "col": 25,
                        "line": 9,
                      },
                      "start": {
                        "col": 8,
                        "line": 9,
                      },
                    },
                    "name": "install",
                    "opts": [],
                    "type": "CommandExpression",
                  },
                ],
                "loc": {
                  "end": {
                    "col": 7,
                    "line": 10,
                  },
                  "start": {
                    "col": 57,
                    "line": 7,
                  },
                },
                "type": "BlockExpression",
              },
            ],
            "loc": {
              "end": {
                "col": 7,
                "line": 10,
              },
              "start": {
                "col": 6,
                "line": 7,
              },
            },
            "name": "connect",
            "opts": [],
            "type": "CommandExpression",
          },
        ],
        "loc": {
          "end": {
            "col": 5,
            "line": 11,
          },
          "start": {
            "col": 58,
            "line": 5,
          },
        },
        "type": "BlockExpression",
      },
    ],
    "loc": {
      "end": {
        "col": 5,
        "line": 11,
      },
      "start": {
        "col": 4,
        "line": 5,
      },
    },
    "module": "ar",
    "name": "connect",
    "opts": [],
    "type": "CommandExpression",
  },
]
`;

exports[`Cas11AST > when fetching commands until a specific line > when given a set of global scope command names > should fetch them correctly 1`] = `
[
  {
    "args": [
      {
        "left": {
          "loc": {
            "end": {
              "col": 17,
              "line": 2,
            },
            "start": {
              "col": 9,
              "line": 2,
            },
          },
          "type": "ProbableIdentifier",
          "value": "aragonos",
        },
        "loc": {
          "end": {
            "col": 23,
            "line": 2,
          },
          "start": {
            "col": 9,
            "line": 2,
          },
        },
        "right": {
          "loc": {
            "end": {
              "col": 23,
              "line": 2,
            },
            "start": {
              "col": 21,
              "line": 2,
            },
          },
          "type": "ProbableIdentifier",
          "value": "ar",
        },
        "type": "AsExpression",
      },
    ],
    "loc": {
      "end": {
        "col": 23,
        "line": 2,
      },
      "start": {
        "col": 4,
        "line": 2,
      },
    },
    "name": "load",
    "opts": [],
    "type": "CommandExpression",
  },
  {
    "args": [
      {
        "left": {
          "loc": {
            "end": {
              "col": 15,
              "line": 3,
            },
            "start": {
              "col": 9,
              "line": 3,
            },
          },
          "type": "ProbableIdentifier",
          "value": "giveth",
        },
        "loc": {
          "end": {
            "col": 22,
            "line": 3,
          },
          "start": {
            "col": 9,
            "line": 3,
          },
        },
        "right": {
          "loc": {
            "end": {
              "col": 22,
              "line": 3,
            },
            "start": {
              "col": 19,
              "line": 3,
            },
          },
          "type": "ProbableIdentifier",
          "value": "giv",
        },
        "type": "AsExpression",
      },
    ],
    "loc": {
      "end": {
        "col": 22,
        "line": 3,
      },
      "start": {
        "col": 4,
        "line": 3,
      },
    },
    "name": "load",
    "opts": [],
    "type": "CommandExpression",
  },
  {
    "args": [
      {
        "loc": {
          "end": {
            "col": 57,
            "line": 5,
          },
          "start": {
            "col": 15,
            "line": 5,
          },
        },
        "type": "AddressLiteral",
        "value": "0x1fc7e8d8e4bbbef77a4d035aec189373b52125a8",
      },
      {
        "body": [
          {
            "args": [
              {
                "loc": {
                  "end": {
                    "col": 23,
                    "line": 6,
                  },
                  "start": {
                    "col": 10,
                    "line": 6,
                  },
                },
                "type": "VariableIdentifier",
                "value": "$dao1Variable",
              },
              {
                "loc": {
                  "end": {
                    "col": 29,
                    "line": 6,
                  },
                  "start": {
                    "col": 24,
                    "line": 6,
                  },
                },
                "type": "ProbableIdentifier",
                "value": "agent",
              },
            ],
            "loc": {
              "end": {
                "col": 29,
                "line": 6,
              },
              "start": {
                "col": 6,
                "line": 6,
              },
            },
            "name": "set",
            "opts": [],
            "type": "CommandExpression",
          },
          {
            "args": [
              {
                "loc": {
                  "end": {
                    "col": 56,
                    "line": 7,
                  },
                  "start": {
                    "col": 14,
                    "line": 7,
                  },
                },
                "type": "AddressLiteral",
                "value": "0x8ccbeab14b5ac4a431fffc39f4bec4089020a155",
              },
              {
                "body": [
                  {
                    "args": [
                      {
                        "loc": {
                          "end": {
                            "col": 25,
                            "line": 8,
                          },
                          "start": {
                            "col": 12,
                            "line": 8,
                          },
                        },
                        "type": "VariableIdentifier",
                        "value": "$dao2Variable",
                      },
                      {
                        "loc": {
                          "end": {
                            "col": 31,
                            "line": 8,
                          },
                          "start": {
                            "col": 26,
                            "line": 8,
                          },
                        },
                        "type": "ProbableIdentifier",
                        "value": "vault",
                      },
                    ],
                    "loc": {
                      "end": {
                        "col": 31,
                        "line": 8,
                      },
                      "start": {
                        "col": 8,
                        "line": 8,
                      },
                    },
                    "name": "set",
                    "opts": [],
                    "type": "CommandExpression",
                  },
                  {
                    "args": [
                      {
                        "loc": {
                          "end": {
                            "col": 25,
                            "line": 9,
                          },
                          "start": {
                            "col": 16,
                            "line": 9,
                          },
                        },
                        "type": "ProbableIdentifier",
                        "value": "vault:new",
                      },
                    ],
                    "loc": {
                      "end": {
                        "col": 25,
                        "line": 9,
                      },
                      "start": {
                        "col": 8,
                        "line": 9,
                      },
                    },
                    "name": "install",
                    "opts": [],
                    "type": "CommandExpression",
                  },
                ],
                "loc": {
                  "end": {
                    "col": 7,
                    "line": 10,
                  },
                  "start": {
                    "col": 57,
                    "line": 7,
                  },
                },
                "type": "BlockExpression",
              },
            ],
            "loc": {
              "end": {
                "col": 7,
                "line": 10,
              },
              "start": {
                "col": 6,
                "line": 7,
              },
            },
            "name": "connect",
            "opts": [],
            "type": "CommandExpression",
          },
        ],
        "loc": {
          "end": {
            "col": 5,
            "line": 11,
          },
          "start": {
            "col": 58,
            "line": 5,
          },
        },
        "type": "BlockExpression",
      },
    ],
    "loc": {
      "end": {
        "col": 5,
        "line": 11,
      },
      "start": {
        "col": 4,
        "line": 5,
      },
    },
    "module": "ar",
    "name": "connect",
    "opts": [],
    "type": "CommandExpression",
  },
  {
    "args": [
      {
        "loc": {
          "end": {
            "col": 23,
            "line": 6,
          },
          "start": {
            "col": 10,
            "line": 6,
          },
        },
        "type": "VariableIdentifier",
        "value": "$dao1Variable",
      },
      {
        "loc": {
          "end": {
            "col": 29,
            "line": 6,
          },
          "start": {
            "col": 24,
            "line": 6,
          },
        },
        "type": "ProbableIdentifier",
        "value": "agent",
      },
    ],
    "loc": {
      "end": {
        "col": 29,
        "line": 6,
      },
      "start": {
        "col": 6,
        "line": 6,
      },
    },
    "name": "set",
    "opts": [],
    "type": "CommandExpression",
  },
  {
    "args": [
      {
        "loc": {
          "end": {
            "col": 56,
            "line": 7,
          },
          "start": {
            "col": 14,
            "line": 7,
          },
        },
        "type": "AddressLiteral",
        "value": "0x8ccbeab14b5ac4a431fffc39f4bec4089020a155",
      },
      {
        "body": [
          {
            "args": [
              {
                "loc": {
                  "end": {
                    "col": 25,
                    "line": 8,
                  },
                  "start": {
                    "col": 12,
                    "line": 8,
                  },
                },
                "type": "VariableIdentifier",
                "value": "$dao2Variable",
              },
              {
                "loc": {
                  "end": {
                    "col": 31,
                    "line": 8,
                  },
                  "start": {
                    "col": 26,
                    "line": 8,
                  },
                },
                "type": "ProbableIdentifier",
                "value": "vault",
              },
            ],
            "loc": {
              "end": {
                "col": 31,
                "line": 8,
              },
              "start": {
                "col": 8,
                "line": 8,
              },
            },
            "name": "set",
            "opts": [],
            "type": "CommandExpression",
          },
          {
            "args": [
              {
                "loc": {
                  "end": {
                    "col": 25,
                    "line": 9,
                  },
                  "start": {
                    "col": 16,
                    "line": 9,
                  },
                },
                "type": "ProbableIdentifier",
                "value": "vault:new",
              },
            ],
            "loc": {
              "end": {
                "col": 25,
                "line": 9,
              },
              "start": {
                "col": 8,
                "line": 9,
              },
            },
            "name": "install",
            "opts": [],
            "type": "CommandExpression",
          },
        ],
        "loc": {
          "end": {
            "col": 7,
            "line": 10,
          },
          "start": {
            "col": 57,
            "line": 7,
          },
        },
        "type": "BlockExpression",
      },
    ],
    "loc": {
      "end": {
        "col": 7,
        "line": 10,
      },
      "start": {
        "col": 6,
        "line": 7,
      },
    },
    "name": "connect",
    "opts": [],
    "type": "CommandExpression",
  },
  {
    "args": [
      {
        "loc": {
          "end": {
            "col": 25,
            "line": 8,
          },
          "start": {
            "col": 12,
            "line": 8,
          },
        },
        "type": "VariableIdentifier",
        "value": "$dao2Variable",
      },
      {
        "loc": {
          "end": {
            "col": 31,
            "line": 8,
          },
          "start": {
            "col": 26,
            "line": 8,
          },
        },
        "type": "ProbableIdentifier",
        "value": "vault",
      },
    ],
    "loc": {
      "end": {
        "col": 31,
        "line": 8,
      },
      "start": {
        "col": 8,
        "line": 8,
      },
    },
    "name": "set",
    "opts": [],
    "type": "CommandExpression",
  },
  {
    "args": [
      {
        "loc": {
          "end": {
            "col": 25,
            "line": 9,
          },
          "start": {
            "col": 16,
            "line": 9,
          },
        },
        "type": "ProbableIdentifier",
        "value": "vault:new",
      },
    ],
    "loc": {
      "end": {
        "col": 25,
        "line": 9,
      },
      "start": {
        "col": 8,
        "line": 9,
      },
    },
    "name": "install",
    "opts": [],
    "type": "CommandExpression",
  },
]
`;

exports[`Cas11AST > when fetching commands until a specific line > when given a set of global scope command names > should fetch them correctly when giving a line higher than the maximum script line 1`] = `
[
  {
    "args": [
      {
        "left": {
          "loc": {
            "end": {
              "col": 17,
              "line": 2,
            },
            "start": {
              "col": 9,
              "line": 2,
            },
          },
          "type": "ProbableIdentifier",
          "value": "aragonos",
        },
        "loc": {
          "end": {
            "col": 23,
            "line": 2,
          },
          "start": {
            "col": 9,
            "line": 2,
          },
        },
        "right": {
          "loc": {
            "end": {
              "col": 23,
              "line": 2,
            },
            "start": {
              "col": 21,
              "line": 2,
            },
          },
          "type": "ProbableIdentifier",
          "value": "ar",
        },
        "type": "AsExpression",
      },
    ],
    "loc": {
      "end": {
        "col": 23,
        "line": 2,
      },
      "start": {
        "col": 4,
        "line": 2,
      },
    },
    "name": "load",
    "opts": [],
    "type": "CommandExpression",
  },
  {
    "args": [
      {
        "left": {
          "loc": {
            "end": {
              "col": 15,
              "line": 3,
            },
            "start": {
              "col": 9,
              "line": 3,
            },
          },
          "type": "ProbableIdentifier",
          "value": "giveth",
        },
        "loc": {
          "end": {
            "col": 22,
            "line": 3,
          },
          "start": {
            "col": 9,
            "line": 3,
          },
        },
        "right": {
          "loc": {
            "end": {
              "col": 22,
              "line": 3,
            },
            "start": {
              "col": 19,
              "line": 3,
            },
          },
          "type": "ProbableIdentifier",
          "value": "giv",
        },
        "type": "AsExpression",
      },
    ],
    "loc": {
      "end": {
        "col": 22,
        "line": 3,
      },
      "start": {
        "col": 4,
        "line": 3,
      },
    },
    "name": "load",
    "opts": [],
    "type": "CommandExpression",
  },
  {
    "args": [
      {
        "loc": {
          "end": {
            "col": 57,
            "line": 5,
          },
          "start": {
            "col": 15,
            "line": 5,
          },
        },
        "type": "AddressLiteral",
        "value": "0x1fc7e8d8e4bbbef77a4d035aec189373b52125a8",
      },
      {
        "body": [
          {
            "args": [
              {
                "loc": {
                  "end": {
                    "col": 23,
                    "line": 6,
                  },
                  "start": {
                    "col": 10,
                    "line": 6,
                  },
                },
                "type": "VariableIdentifier",
                "value": "$dao1Variable",
              },
              {
                "loc": {
                  "end": {
                    "col": 29,
                    "line": 6,
                  },
                  "start": {
                    "col": 24,
                    "line": 6,
                  },
                },
                "type": "ProbableIdentifier",
                "value": "agent",
              },
            ],
            "loc": {
              "end": {
                "col": 29,
                "line": 6,
              },
              "start": {
                "col": 6,
                "line": 6,
              },
            },
            "name": "set",
            "opts": [],
            "type": "CommandExpression",
          },
          {
            "args": [
              {
                "loc": {
                  "end": {
                    "col": 56,
                    "line": 7,
                  },
                  "start": {
                    "col": 14,
                    "line": 7,
                  },
                },
                "type": "AddressLiteral",
                "value": "0x8ccbeab14b5ac4a431fffc39f4bec4089020a155",
              },
              {
                "body": [
                  {
                    "args": [
                      {
                        "loc": {
                          "end": {
                            "col": 25,
                            "line": 8,
                          },
                          "start": {
                            "col": 12,
                            "line": 8,
                          },
                        },
                        "type": "VariableIdentifier",
                        "value": "$dao2Variable",
                      },
                      {
                        "loc": {
                          "end": {
                            "col": 31,
                            "line": 8,
                          },
                          "start": {
                            "col": 26,
                            "line": 8,
                          },
                        },
                        "type": "ProbableIdentifier",
                        "value": "vault",
                      },
                    ],
                    "loc": {
                      "end": {
                        "col": 31,
                        "line": 8,
                      },
                      "start": {
                        "col": 8,
                        "line": 8,
                      },
                    },
                    "name": "set",
                    "opts": [],
                    "type": "CommandExpression",
                  },
                  {
                    "args": [
                      {
                        "loc": {
                          "end": {
                            "col": 25,
                            "line": 9,
                          },
                          "start": {
                            "col": 16,
                            "line": 9,
                          },
                        },
                        "type": "ProbableIdentifier",
                        "value": "vault:new",
                      },
                    ],
                    "loc": {
                      "end": {
                        "col": 25,
                        "line": 9,
                      },
                      "start": {
                        "col": 8,
                        "line": 9,
                      },
                    },
                    "name": "install",
                    "opts": [],
                    "type": "CommandExpression",
                  },
                ],
                "loc": {
                  "end": {
                    "col": 7,
                    "line": 10,
                  },
                  "start": {
                    "col": 57,
                    "line": 7,
                  },
                },
                "type": "BlockExpression",
              },
            ],
            "loc": {
              "end": {
                "col": 7,
                "line": 10,
              },
              "start": {
                "col": 6,
                "line": 7,
              },
            },
            "name": "connect",
            "opts": [],
            "type": "CommandExpression",
          },
        ],
        "loc": {
          "end": {
            "col": 5,
            "line": 11,
          },
          "start": {
            "col": 58,
            "line": 5,
          },
        },
        "type": "BlockExpression",
      },
    ],
    "loc": {
      "end": {
        "col": 5,
        "line": 11,
      },
      "start": {
        "col": 4,
        "line": 5,
      },
    },
    "module": "ar",
    "name": "connect",
    "opts": [],
    "type": "CommandExpression",
  },
  {
    "args": [
      {
        "loc": {
          "end": {
            "col": 23,
            "line": 6,
          },
          "start": {
            "col": 10,
            "line": 6,
          },
        },
        "type": "VariableIdentifier",
        "value": "$dao1Variable",
      },
      {
        "loc": {
          "end": {
            "col": 29,
            "line": 6,
          },
          "start": {
            "col": 24,
            "line": 6,
          },
        },
        "type": "ProbableIdentifier",
        "value": "agent",
      },
    ],
    "loc": {
      "end": {
        "col": 29,
        "line": 6,
      },
      "start": {
        "col": 6,
        "line": 6,
      },
    },
    "name": "set",
    "opts": [],
    "type": "CommandExpression",
  },
  {
    "args": [
      {
        "loc": {
          "end": {
            "col": 25,
            "line": 8,
          },
          "start": {
            "col": 12,
            "line": 8,
          },
        },
        "type": "VariableIdentifier",
        "value": "$dao2Variable",
      },
      {
        "loc": {
          "end": {
            "col": 31,
            "line": 8,
          },
          "start": {
            "col": 26,
            "line": 8,
          },
        },
        "type": "ProbableIdentifier",
        "value": "vault",
      },
    ],
    "loc": {
      "end": {
        "col": 31,
        "line": 8,
      },
      "start": {
        "col": 8,
        "line": 8,
      },
    },
    "name": "set",
    "opts": [],
    "type": "CommandExpression",
  },
  {
    "args": [
      {
        "body": [
          {
            "args": [
              {
                "loc": {
                  "end": {
                    "col": 19,
                    "line": 14,
                  },
                  "start": {
                    "col": 13,
                    "line": 14,
                  },
                },
                "type": "ProbableIdentifier",
                "value": "voting",
              },
              {
                "loc": {
                  "end": {
                    "col": 33,
                    "line": 14,
                  },
                  "start": {
                    "col": 20,
                    "line": 14,
                  },
                },
                "type": "ProbableIdentifier",
                "value": "token-manager",
              },
              {
                "loc": {
                  "end": {
                    "col": 43,
                    "line": 14,
                  },
                  "start": {
                    "col": 34,
                    "line": 14,
                  },
                },
                "type": "ProbableIdentifier",
                "value": "MINT_ROLE",
              },
            ],
            "loc": {
              "end": {
                "col": 43,
                "line": 14,
              },
              "start": {
                "col": 6,
                "line": 14,
              },
            },
            "name": "revoke",
            "opts": [],
            "type": "CommandExpression",
          },
        ],
        "loc": {
          "end": {
            "col": 5,
            "line": 15,
          },
          "start": {
            "col": 31,
            "line": 13,
          },
        },
        "type": "BlockExpression",
      },
    ],
    "loc": {
      "end": {
        "col": 5,
        "line": 15,
      },
      "start": {
        "col": 4,
        "line": 13,
      },
    },
    "module": "ar",
    "name": "connect",
    "opts": [],
    "type": "CommandExpression",
  },
  {
    "args": [
      {
        "loc": {
          "end": {
            "col": 28,
            "line": 17,
          },
          "start": {
            "col": 8,
            "line": 17,
          },
        },
        "type": "VariableIdentifier",
        "value": "$globalScopeVariable",
      },
      {
        "loc": {
          "end": {
            "col": 35,
            "line": 17,
          },
          "start": {
            "col": 29,
            "line": 17,
          },
        },
        "type": "StringLiteral",
        "value": "test",
      },
    ],
    "loc": {
      "end": {
        "col": 35,
        "line": 17,
      },
      "start": {
        "col": 4,
        "line": 17,
      },
    },
    "name": "set",
    "opts": [],
    "type": "CommandExpression",
  },
]
`;
